### PR TITLE
refactor navier_stokes_calculators

### PR DESCRIPTION
**Pull request by richardschu**
Refactor the `navier_stokes_calculators` (#348 was closed accidentally pushing without commits to the branch, when trying to squash the commits. Sorry about that.)

### CODE REVIEW

File: include/exadg/operators/navier_stokes_calculators.cpp [hunk 1]
```diff
@@ -23,6 +23,66 @@
 
 namespace ExaDG
 {
+template<int dim, typename Number>
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.cpp:26]: this was just copied

File: include/exadg/operators/navier_stokes_calculators.cpp [hunk 2]
```diff
@@ -23,6 +23,66 @@
...
+DivergenceCalculator<dim, Number>::DivergenceCalculator()
+  : matrix_free(nullptr), dof_index_vector(1), dof_index_scalar(2), quad_index(0)
+{
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.cpp:29]: this was just copied

File: include/exadg/operators/navier_stokes_calculators.cpp [hunk 3]
```diff
@@ -84,10 +144,232 @@ ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> cons
...
+  : matrix_free(nullptr), dof_index(0), quad_index(0)
+{
+}
+
+template<int dim, typename Number>
+void
+VorticityCalculator<dim, Number>::initialize(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+                                             unsigned int const                      dof_index_in,
+                                             unsigned int const                      quad_index_in)
+{
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.cpp:158]: this was just copied as well

File: include/exadg/operators/navier_stokes_calculators.cpp [hunk 4]
```diff
@@ -84,10 +144,232 @@ ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> cons
...
+
+template<int dim, typename Number>
+QCriterionCalculator<dim, Number>::QCriterionCalculator()
+  : matrix_free(nullptr),
+    dof_index_u(0),
+    dof_index_u_scalar(0),
+    quad_index(0),
+    compressible_flow(false)
+{
+}
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.cpp:269]: compressible_flow now as member variable

File: include/exadg/operators/navier_stokes_calculators.cpp [hunk 5]
```diff
@@ -84,10 +144,232 @@ ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> cons
...
+void
+QCriterionCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                             VectorType &                            dst,
+                                             VectorType const &                      src,
+                                             Range const & cell_range) const
+{
+  CellIntegratorVector integrator_vector(matrix_free, dof_index_u, quad_index);
+  CellIntegratorScalar integrator_scalar(matrix_free, dof_index_u_scalar, quad_index);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.cpp:306]: if statement before the loop or inside? minor code duplication or not?

File: include/exadg/operators/navier_stokes_calculators.h [hunk 6]
```diff
@@ -30,6 +30,43 @@
 
 namespace ExaDG
 {
+template<int dim, typename Number>
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.h:33]: copied as well

File: include/exadg/operators/navier_stokes_calculators.h [hunk 7]
```diff
@@ -72,6 +109,127 @@ class ShearRateCalculator
   unsigned int quad_index;
 };
 
+template<int dim, typename Number>
```
Comment (richardschu) [inline include/exadg/operators/navier_stokes_calculators.h:112]: copied as well

File: include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp [hunk 8]
```diff
@@ -1197,7 +1198,7 @@ void
 SpatialOperatorBase<dim, Number>::compute_vorticity_magnitude(VectorType &       dst,
                                                               VectorType const & src) const
 {
-  velocity_magnitude_calculator.compute(dst, src);
+  magnitude_calculator.compute(dst, src);
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp:1201]: thanks for this general improvement! This has caused confusion in the past.

File: include/exadg/operators/navier_stokes_calculators.cpp (hunk 2 above)
Comment (nfehn) [reply include/exadg/operators/navier_stokes_calculators.cpp:29]: what about the values 1 and 2 here? The shear rate calculator uses 0 always for initialization.

File: include/exadg/operators/navier_stokes_calculators.cpp (hunk 5 above)
Comment (nfehn) [reply include/exadg/operators/navier_stokes_calculators.cpp:306]: i agree with the present version

**Comment by nfehn** [comment]
I would suggest to unify the default dof indices. Apart from that, we are ready to merge from my side.

File: include/exadg/operators/navier_stokes_calculators.cpp (hunk 2 above)
Comment (richardschu) [reply include/exadg/operators/navier_stokes_calculators.cpp:29]: I changed all of these to 0. They were set in the constructor anyways ... but now its uniformly initialized to 0.

**Review by kronbichler — APPROVED** [review]
Nice simplification.

**Review by nfehn — APPROVED** [review]

**Pull request merged**